lib/src/suggestor_mixins.dart

```diff
@@ -0,0 +1,114 @@
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/element/element.dart';
+import 'package:codemod/src/file_context.dart';
+
+import 'patch.dart';
+import 'suggestor.dart';
+
+/// Mixin that implements the [Suggestor] interface and makes it easier to write
+/// suggestors that operate as an [AstVisitor].
+///
+/// With the [AstVisitor] pattern, you can override the applicable `visit`
+/// methods to find what you're looking for and generate patches at specific
+/// locations in the source using the offsets provided by the [AstNode]s and
+/// tokens therein.
+///
+/// Note that this mixin provides an implementation of [generatePatches] that
+/// should not need to be overridden except for performance optimization reasons
+/// like avoiding analysis on certain files.
+///
+/// By default, this operates on the unresolved AST. Subclasses that need a
+/// fully resolved AST (e.g. for static typing info) should override
+/// [shouldResolveAst] to return true.
+///
+/// The easiest way to understand this pattern is to see an example. Consider
+/// the following suggestor that aims to remove all deprecated declarations:
+///
+///     import 'package:analyzer/analyzer.dart';
+///     import 'package:codemod/codemod.dart';
+///
+///     class DeprecatedRemover extends GeneralizingAstVisitor
+///         with UnresolvedAstVisitingSuggestor {
+///
+///       static bool isDeprecated(AnnotatedNode node) =>
+///           node.metadata.any((m) => m.name.name.toLowerCase() == 'deprecated');
+///
+///       @override
+///       visitDeclaration(Declaration node) {
+///         if (isDeprecated(node)) {
+///           // Remove the node by replacing the span from its start offset to its end
+///           // offset with an empty string.
+///           yieldPatch('', node.offset, node.end);
+///         }
+///       }
+///     }
+mixin AstVisitingSuggestor<R> on AstVisitor<R> implements Suggestor {
+  final _patches = <Patch>{};
+
+  /// The context helper for the file for which patches are currently being
+  /// generated.
+  FileContext get context => _context;
+  FileContext _context;
+
+  @override
+  Stream<Patch> generatePatches(FileContext context) async* {
+    _patches.clear();
```

Comment:
Since this method relies on _patches and _context, there's a possibility a race condition occurring if `generatePatches` is called a second time on the same object before the stream from the first call has closed.

(These cases are mostly written out to make sure I was thinking everything through properly 😄 ) If the second call happens:
1. before patches are yielded, the wrong context would be used and _patches would be shared for files
2. after the creation of `Stream.fromIterable(_patches)` but before it's actually iterated, `_patches` will be cleared, which will result in no patches getting yielded
3. after the creation of `Stream.fromIterable(_patches)` but while it's being iterated,  `_patches` will be cleared, which would result in a concurrent iteration error being thrown

I'm not sure if it'd be possible for this to occur during normal usage, if it does occur it could cause silent failures or be hard to track down.

We could set a flag and throw to prevent this:
```dart
Stream<Patch> generatePatches(FileContext context) async* {
  if (_isGeneratingPatches) {
    throw StateError('Must wait for other generatePatches calls to complete');
  }

  try {
    _isGeneratingPatches = true;
    ...
  } finally {
    _isGeneratingPatches = false;
  }
}
```
or potentially use something like package:synchronized's locks to defer subsequent calls until current calls complete.

Thoughts?

Reply:
Hmm, yeah this is something that has sort of bugged me about how the visitor mixins are implemented right now. When I was playing around with this refactor, a couple of ideas I considered were:

1. Using a "provider" pattern where a provider is given to `runInteractiveCodemod` and that provider provides a `Suggestor` (could be a class or a function) that is only ever used once.
1. Making `Suggestor` a typedef instead of a class so that each invocation could be run and scoped independently

The provider pattern seemed a bit overkill, so I wanted to avoid it (why force authors to manage an extra layer of indirection when a single, simple class can work).

And the fact that AST/Element visitors are implemented as classes means it is convenient to be able to define a single class that is both the visitor and the suggestor, but as you point out, it forces us to deal with the problem of storing state on the class that is shared across all input files.

---

After I wrote the above I started to think that maybe changing `Suggestor` to a typedef would help with this, so I gave it a shot: https://github.com/Workiva/dart_codemod/pull/40/files

I think I really like how this turned out. Suggestors can be written as really simple functions, which is nice, but can also be written as classes that are callable by overriding `Stream<Patch> call(FileContext context)`. That makes it possible to continue using a visitor directly as a suggestor like we do now.

I also refactored `Patch` a bit so that the publicly exported class is very simple (updatedText, startOffset, endOffset), and the rest of the logic needed to actually apply patches is moved to `SourcePatch` which is kept private. This allowed me to get rid of the `FileContext.patch()` util so that consumers can just `yield Patch(...)` directly (bonus: this aligns nicely with the `yieldPatch()` method on our visitor mixins), and the visitor mixins no longer need a reference to the `SourceFile`, which gets rid of the awkward class state to hold the currently active `FileContext`.

Let me know what you think!

Reply:
I like the typedef version, but don't have a strong preference either way.

Reply:
I like it as well!! ✨ 

(I left a couple small comments on that PR too)